### PR TITLE
Feaure matching API

### PR DESCRIPTION
**Pull request by theNded**
- Expose `correspondences from features` api for global registration from external point-wise 3D features (e.g. FCGF).
- Improve FPFH feature matching time:

| Legacy | Tensor (CPU) | Tensor (CUDA) |
|---          |---                    |---                       |
|0.023     | 0.016             | 0.0023               |

------

In theory a tensor RANSAC API could be included as well, but there are various problems. Problem analysis step by step.

1. Correctness: combination check all passed (L for legacy, T for tensor), no problem here:

| FPFH feature | Feature matching | RANSAC | Status  |
|---                    |---                          |---             |---                             |
| L                     | L                          | L              |:heavy_check_mark:|
| L                     | T                          | L              |:heavy_check_mark:|
| T                     | L                          | L              |:heavy_check_mark:|
| T                     | T                          | L              |:heavy_check_mark:|
| T                     | T                          | T              |:heavy_check_mark:|

2. Performance: tensor API is much slower.

Total registration time (s) (100K iterations)

| Legacy | Tensor (CPU) | Tensor (CUDA) |
|---          |---                     |---                      |
|0.088     | 36.04                      |35.74                 |

Average run time breakdown (model selection/estimation/check happens every iteration; model validation only happens 250 times per 100K iterations)
|    | Legacy| Tensor (CPU)| Tensor (CUDA)|
|---|---|---|---|
|Model selection| 10^-7   | 0.002    | 0.0020 |
|Model estimation|  2*10^-6| 0.002 | 0.0022|
|Model check| 10^-7 | 0.0012 | 0.0012|
|Model validation| 0.003 | 0.0024| 0.0045|

Problem:
- **Model selection/estimation/check**: Tensor point cloud operations (`ComputeTransformation`; `SelectByIndex`) are only suitable for large inputs. The tensor overhead is overwhelmingly large for small data.
- **Model validation**: Hybrid NNS isn't faster for 3D on CUDA

Potential solutions:
- **This PR:** Expose feature matching API first and use the legacy RANSAC solver. **Tensor RANSAC is temporarily removed and archived in `wei/ransac-tensor`**
- **Future work:** Optimized (batch) operations for small point clouds, should consist of several PRs: batch SVD for transformation estimation; batch matrix multiplication for transformation; random tensor generation for model selection.

## Type

-   [ ] Bug fix (non-breaking change which fixes an issue): Fixes #
-   [x] New feature (non-breaking change which adds functionality). Resolves #
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) Resolves #

## Checklist:

-   [x] I have run `python util/check_style.py --apply` to apply Open3D **code style**
    to my code.
-   [ ] This PR changes Open3D behavior or adds new functionality.
    -   [ ] Both C++ (Doxygen) and Python (Sphinx / Google style) **documentation** is
        updated accordingly.
    -   [ ] I have added or updated C++ and / or Python **unit tests** OR included **test
        results** (e.g. screenshots or numbers) here.
-   [ ] I will follow up and update the code if CI fails.
    
-   [ ] For fork PRs, I have selected **Allow edits from maintainers**.

## Description

- - -
This change is [Reviewable](https://reviewable.io/reviews/isl-org/Open3D/5964)

### CODE REVIEW

File: cpp/open3d/pipelines/registration/Feature.h [hunk 1]
```diff
@@ -54,6 +56,30 @@ std::shared_ptr<Feature> ComputeFPFHFeature(
...
                 geometry::KDTreeSearchParamKNN());
 
+/// \brief Function to find correspondences via 1-nearest neighbor feature
+/// matching. Target is used to construct a nearest neighbor search
+/// object, in order to query source.
+/// \param source_features (D, N) feature
+/// \param target_features (D, M) feature
+/// \param mutual_filter Boolean flag, only return correspondences s.t. i and j
+/// are mutually the nearest neighbor
+/// \param mutual_filter Boolean flag, only return correspondences (i, j) s.t.
```
Comment (yuecideng) [inline cpp/open3d/pipelines/registration/Feature.h:66]: Duplicate param `mutual_filter`.

File: cpp/open3d/t/pipelines/registration/Feature.cpp [hunk 2]
```diff
@@ -90,6 +90,52 @@ core::Tensor ComputeFPFHFeature(const geometry::PointCloud &input,
...
+                                 corres_ij.GetDtype(), corres_ij.GetDevice());
+
+    // Change view for the appending axis
+    core::Tensor result_ij =
+            arange_source.View({-1, 1}).Append(corres_ij.View({-1, 1}), 1);
+
+    if (!mutual_filter) {
+        return result_ij;
+    }
+
```
Comment (yuecideng) [inline cpp/open3d/t/pipelines/registration/Feature.cpp:114]: Could we use 2 threads to run mutual KnnSearch when `mutual_filter` is on?

**Review by yuecideng — APPROVED** [review]

**Comment by ShaoxiongYao** [comment]
Thank you for providing this pull request! If I understand correctly, this will enable users to get correspondences from a set of features using the `correspondences_from_features`, right? With these features, I may be able to implement non-rigid ICP with my customized deformation solver.

**Comment by theNded** [comment]
> Thank you for providing this pull request! If I understand correctly, this will enable users to get correspondences from a set of features using the `correspondences_from_features`, right? With these features, I may be able to implement non-rigid ICP with my customized deformation solver.

Yes, see
https://github.com/isl-org/Open3D/pull/5964/files#diff-6bdcf782f0d76f4eb437bd89ada11445e23a5008633f09131ab9edf8c5806998R118
CUDA acceleration is also supported with the tensor interface, I should update the example in a bit.

File: cpp/open3d/t/pipelines/registration/Feature.h [hunk 3]
```diff
@@ -37,6 +37,29 @@ core::Tensor ComputeFPFHFeature(
...
+/// \param mutual_consistency_ratio Float threshold to decide whether the number
+/// of correspondences is sufficient. Only used when mutual_filter is set to
+/// True.
+/// \return (K, 2, Int64) tensor. When mutual_filter is disabled: the first
+/// column is arange(0, N) of source, and the second column is the corresponding
+/// index of target. When mutual_filter is enabled, return the filtering subset
+/// of the aforementioned correspondence set where source[i] and target[j] are
+/// mutually the nearest neighbor. If the subset size is smaller than
+/// mutual_consistency_ratio * N, return the unfiltered set.
+core::Tensor CorrespondencesFromFeatures(const core::Tensor &source_features,
```
[the remark this answers is not in the document]
Comment (theNded) [reply cpp/open3d/t/pipelines/registration/Feature.h:56]: Sounds reasonable, will do in a future PR

File: examples/python/pipelines/registration_ransac.py [hunk 4]
```diff
@@ -18,82 +31,119 @@ def preprocess_point_cloud(pcd, voxel_size):
...
+        # shape: (feature_dim, num_features)
+        src_fpfh_np = np.asarray(src_fpfh.data).copy()
+        dst_fpfh_np = np.asarray(dst_fpfh.data).copy()
+
+        src_fpfh_import = o3d.pipelines.registration.Feature()
+        src_fpfh_import.data = src_fpfh_np
+
+        dst_fpfh_import = o3d.pipelines.registration.Feature()
+        dst_fpfh_import.data = dst_fpfh_np
+
```
[the remark this answers is not in the document]
Comment (theNded) [reply examples/python/pipelines/registration_ransac.py:129]: I intend to extend this example with support to this functionality. However, there are inconsistencies between tensor point cloud downsample (nearest corner per voxel) and the legacy (mean per voxel). I will make another PR regarding t point cloud, and revisit this example.

File: cpp/open3d/t/pipelines/registration/Feature.cpp [hunk 5]
```diff
@@ -90,6 +90,60 @@ core::Tensor ComputeFPFHFeature(const geometry::PointCloud &input,
     return fpfh;
 }
 
+core::Tensor CorrespondencesFromFeatures(const core::Tensor &source_features,
```
[the remark this answers is not in the document]
Comment (theNded) [reply cpp/open3d/t/pipelines/registration/Feature.cpp:94]: I think there is no need, as tensor nns backend is able to query properly (and fast). The conversion only has to be done for calling ransac.

**Review by ssheorey — APPROVED** [review]

**Comment by guyuezuntinggithub** [comment]
very great work! very appreciate for this.
can I use tensor FPFH in C++ API?

**Pull request merged**